### PR TITLE
Build and compile against multiple Alfresco versions

### LINKED ISSUE #149: Dynamic Extensions should compile against Alfresco 5.2.g
To ensure compability with the latest alfresco versions, Dynamic extensions should compile against Alfresco 5.2.g

### PR DESCRIPTION
**Pull request by kerkhofsd**
Compile & build against multiple Alfresco versions.

## Description
To be sure the DE works for a specific Alfresco version, is should compile and build against that specific Alfresco version. 
This PR refactors DE so that it creates different artifacts for multiple Alfresco versions.
For each DE artifact (`artifactName` e.g. `annotations`) there now are artifacts for each supported Alfresco major.minor version (`artifactName-alfrescoVersion` e.g. `annotations-52`).

## Related Issue
However this was not the main purpose, fixes #149 

## Motivation and Context
Assets:
- We are sure that DE compiles against the Alfresco / Spring version
- Breaking changes that we need to introduce between multiple Alfresco versions can now be easily solved. Each artifact for a specific Alfresco version can have it's own implementation of classes and resources. The bulk of the code base (non breaking) stays shared.
- Big step towards Alfresco 6 support

Drawbacks:
- Projects depending on DE should now include the correct artifacts, including the Alfresco major minor version.
- All subprojects are split into multiple projects for the corresponding Alfresco versions. This shouldn't be needed for submodules that don't have any Spring or Alfresco dependency. Also we need to check if we could improve the Gradle plugin according to this new behavior.

## How Has This Been Tested?
Manually tested on Alfresco 4.2 & 5.2.4

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have updated the changelog.

### CODE REVIEW

File: alfresco-dynamic-extensions-repo/alfresco-dynamic-extensions-repo-50/dummy.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+To make sure this directory is in git
```
Comment (yregaieg) [inline alfresco-dynamic-extensions-repo/alfresco-dynamic-extensions-repo-50/dummy.txt]: Shouldn't this be an empty file named .gitkeep ?

Comment (kerkhofsd) [reply alfresco-dynamic-extensions-repo/alfresco-dynamic-extensions-repo-50/dummy.txt]: These files are removed in a later commit

Comment (yregaieg) [reply alfresco-dynamic-extensions-repo/alfresco-dynamic-extensions-repo-50/dummy.txt]: Yeah, noticed that already, more of a global remark

**Comment by yregaieg** [comment]
Looks to me not all supported Alfresco versions are supported by the current state of PR, maybe worth making few extra tests around scheduled jobs, json serialization/mapping just to make sure everything works as expected in ACS 6.X

**Comment by kerkhofsd** [comment]
> Looks to me not all supported Alfresco versions are supported by the current state of PR, maybe worth making few extra tests around scheduled jobs, json serialization/mapping just to make sure everything works as expected in ACS 6.X

For now, we don't support ACS 6.X yet. 
However the subprojects names don't imply this, all versions < 6 are supported. I don't have BOM's yet for Alfresco 4.2 & 5.1, that's why I did not yet compile & build against these versions.

Alfresco 6 support will follow.
A lot of our dependencies are not compatible with Alfresco 6 (Spring 5). The purpose of this PR is easily build multiple DE versions with different dependencies (e.g. Gemini Blueprint)

**Comment by stanmine** [comment]
i would restructure the project a little but.

When you want to build your code using DE, you only need the annotations library. All your Alfresco project should do is pull in the annotations jar and perhaps the DE plugin to complete the manifest file and all should be fine. This library is independent from alfresco version, hence there is no need for a separate module for each alfresco version.

It might also be nice (but open for discussion) to use a flavor for a specific alfresco version instead of a custom artifact for each alfresco version. Take a look at [the google guava versioning shema](https://github.com/google/guava/wiki/ReleasePolicy) for example. This makes swapping DE versions in your project easier if you put the version in a property.

**Comment by kerkhofsd** [comment]
Thanks for your feedback @stanmine.

> i would restructure the project a little but.
You mean keep a single artifact for the `annotations` library?

The problem is that the `annotations` library currently does have Alfresco & Spring dependencies. We'll have to take a look how we can remove these dependencies.

**Comment by stanmine** [comment]
Yes, that is indeed what i mean.

It is a great opportunity to refactor :-)

**Comment by kerkhofsd** [comment]
@stanmine 
Went back to a single build artifact for the `annotations` subproject!

File: README.md [hunk 2]
```diff
@@ -33,6 +50,7 @@ Dynamic Extensions is also known to work on:
 * Alfresco Enterprise 4.0
 * Alfresco Enterprise 5.0
 * Alfresco Enterprise 5.1
+* Alfresco Enterprise 5.2
```
Comment (vierbergenlars) [inline README.md]: Tests are run for Alfresco Community 5.2, so maybe that version should be added to _Dynamic Extensions is developed and tested against_

File: alfresco-dynamic-extensions-repo/build.gradle [hunk 3]
```diff
@@ -1,91 +1,107 @@
...
 
-    from(configurations.bundles) {
-        into("config/alfresco/module/$project.name/standard-bundles")
-    }
+    def alfrescoAmp = file("${file("$buildDir/amps")}/${project.name}-${project.version}.amp")
 
-    from(configurations.amp) {
-        into("lib")
+    artifacts {
+        archives sourceJar, javadocJar, alfrescoAmp
```
Comment (vierbergenlars) [inline alfresco-dynamic-extensions-repo/build.gradle]: Referring directly to a file in `artifacts {}` is not recommended.
You can depend directly on the `amp` task if you move the `amp` task definition above artifacts

File: gradle-plugin/src/main/groovy/com/github/dynamicextensionsalfresco/gradle/DynamicExtensionPlugin.groovy [hunk 4]
```diff
@@ -45,15 +45,18 @@ class DynamicExtensionPlugin implements Plugin<Project> {
 		def dynamicExtensions = [
 			version: project.alfrescoDynamicExtensions.versions.dynamicExtensions
 		]
+		def alfrescoVersion = surf.version.toString().replace('.', '')substring(0, 2)
```
Comment (vierbergenlars) [inline gradle-plugin/src/main/groovy/com/github/dynamicextensionsalfresco/gradle/DynamicExtensionPlugin.groovy]: Missing a `.` before the call to `substring`

**Comment by kerkhofsd** [comment]
Thanks for the feedback @vierbergenlars. 
See update...

**Pull request merged**